### ISSUE TITLE
Improve build system

### ISSUE THREAD

**Issue by mortendahl** [tf-encrypted/tf-encrypted#548]
The current build system has some limitations, including scalability since everything is in one file.

- [ ] move to bazel?
- [ ] split make files into subdirectories
- [ ] package up more compiled custom ops

**Comment by justin1121** [comment]
Bazel examples from Tensorflow: https://github.com/tensorflow/custom-op
Bazel build I've created: https://github.com/dropoutlabs/tf-trusted/blob/master/tf_trusted_custom_op/BUILD

I think its pretty smooth!

**Comment by justin1121** [comment]
Check out https://github.com/tf-encrypted/tf-encrypted/pull/560 for an example of how bazel would be set up. You then can run the NTL test with `bazel build ntl:ntl_ops_py_test && bazel-bin/ntl/ntl_ops_py_test`

**Comment by ianlivingstone** [comment]
Moving to bazel would also mean replacing our publishing functionality. If we're considering moving, let's do a bit of a pro/con analysis on the different approaches. If we do switch to bazel it should be a wholesale switch meaning all of the functionality in the Makefile would need to be ported.

**Comment by justin1121** [comment]
> If we do switch to bazel it should be a wholesale switch meaning all of the functionality in the Makefile would need to be ported.

I don't agree with this and would definitely make me not want to use bazel at all. Using bazel to build the custom operations and make to do everything else seems fine to me.

**Comment by ianlivingstone** [comment]
> I don't agree with this and would definitely make me not want to use bazel at all. Using bazel to build the custom operations and make to do everything else seems fine to me.

Can you explain why the pros and cons of either approach, I'd like to understand why you wouldn't want to go this route?

**Comment by justin1121** [comment]
Sure. I'll do the pros and cons of `make` vs `bazel`.

### Make

**Pros**

- We already have an alright build system with make
- Stable
- Flexible
- Lots of documentation and examples
- Easily interop with projects that also use `make`
- Most systems come with make or is super easy to install

**Cons**

- Tricky to understand
- Tricky to debug
- Tricky to write Makefiles
- Annoying to use for compiling complex C++ projects with lots of files and dependencies

### Bazel

**Pros**

- Can make compiling complex C++ projects easier, especially if the dependencies also use `bazel`
- Flexible
- Easily interop with projects that also use `bazel`

**Cons**

- Documentation/examples lacking
- Tough to interop with projects that use `make`
- Tricky to understand
- Tricky to debug
- Extra thing users would have to install, can be a little tricky to install
- Not stable, every new version seems to break things

I'm sure there are others but this is what I have so far!

### Discussion

One of the greatest strengths of `bazel` I think is its ability to interop with other projects (dependencies) that also use `bazel`. But, this breaks down pretty quickly when you run into a project that doesn't use `bazel`. To get a project to that doesn't use `bazel` to work with `bazel` you can do one of three things:

1. Write `bazel` `BUILD` file for that dependency that basically rewrites the build system of the dependency.
2. Use https://github.com/bazelbuild/rules_foreign_cc to import non-bazel projects (just found this haven't actually tried this out).
3. Build dependencies with `make` and point `bazel` to the installed headers/libs. This is what I've done for NTL Poc (https://github.com/tf-encrypted/tf-encrypted/pull/560).

Some of bazels other weaknesses are that it breaks backwards compatibility a lot (unstable) and the documentation and examples are either hard to understand or non-existent.

One other thing that's tricky with a full switch to `bazel` is that we're using `make` a lot for building pip packages and docker builds. It looks like there is some support for building docker images (https://github.com/bazelbuild/rules_docker) but its just another thing we have to learn and I'm not sure what we're getting out of that. As for pip packages even tensorflow uses just a script to build its pip package (https://github.com/tensorflow/custom-op/blob/master/build_pip_pkg.sh), which is probably fine but its not directly using `bazel`.

All of the above basically makes me think that we should either not use bazel or only use it for building custom operations. I'm not even sold on using it for only custom operations if interop with non-bazel projects is better with https://github.com/bazelbuild/rules_foreign_cc.

@ianlivingstone would be great if you could explain why you think the wholesale switch to bazel would be necessary.

**Comment by ianlivingstone** [comment]
>@ianlivingstone would be great if you could explain why you think the wholesale switch to bazel would be necessary.

My thought process was coming from a place of simplicity: reducing the number of moving parts and things someone needs to understand to contribute as much as possible.

Given your above research and discussion, I'm concerned that bazel doesn't solve the problem sufficiently well enough versus a well-crafted Makefile system. We should concretely identify the problems we're trying to solve: is it simply the custom op build workflow, or is there more underneath?

It would also help to identify the framework of a good solution (e.g. what do we want/need in a solution).

**Comment by justin1121** [comment]
> is it simply the custom op build workflow, or is there more underneath?

I was definitely just thinking it would help the custom op build workflow. At this point I'm definitely leaning towards just using `make` and we can figure out how to clean up our build system a little bit. Separate the makefile into smaller bit etc.

Would be great if @mortendahl could give some ideas as well around this.

**Comment by mortendahl** [comment]
I like `cargo build`

**Comment by mortendahl** [comment]
Great run down @justin1121! I would be inclined to stay with Makefiles for now, but find a way to isolate concerns if possible, such that building eg secure randomness doesn't happen in the same file as pushing to pypi.

Alternatively, or if it keeps being a problem, it sounds to me that a mix between bazel and Makefiles could work as well.

**Comment by mortendahl** [comment]
@justin1121 given what we know now, has anything changed in your mind re moving to bazel?

**Comment by justin1121** [comment]
Wouldn't say anything has changed from my above pros/cons list and discussion above!
